### PR TITLE
Migrate from structopt to clap

### LINKED ISSUE #2365: Migrate from `structopt` to `clap`
**Issue summary**

Quote from [`structopt` doc](https://github.com/TeXitoi/structopt)
> As clap v3 is now out, and the structopt features are integrated into (almost as-is), structopt is now in maintenance mode: no new feature will be added.
Bugs will be fixed, and documentation improvements will be accepted.
See the [structopt -> clap migration guide](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#migrating)

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
## Summary of changes

Changes introduced in this pull request:
- 

## Reference issue to close (if applicable)

Closes #2365

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: forest/shared/src/cli/mod.rs [hunk 1]
```diff
@@ -20,97 +21,92 @@ use once_cell::sync::Lazy;
...
-    #[structopt(long)]
+    #[arg(long)]
     pub encrypt_keystore: Option<bool>,
     /// Choose network chain to sync to
-    #[structopt(
-        long,
-        default_value = "mainnet",
-        possible_values = &["mainnet", "calibnet"],
-    )]
+    #[arg(long, default_value = "mainnet")]
```
Comment (creativcoder) [inline forest/shared/src/cli/mod.rs:85]: does clap not have something to support "possible_values"?

File: utils/statediff/src/main.rs [hunk 2]
```diff
@@ -1,68 +1,86 @@
...
-    post: Cid,
-    /// The name of the chain
-    #[structopt(short, long, default_value = "mainnet")]
-    chain: String,
-    /// The depth at which IPLD links are resolved
-    #[structopt(short, long)]
-    depth: Option<u64>,
-}
+// /// Examine the state delta
+// #[derive(clap::Subcommand)]
```
Comment (creativcoder) [inline utils/statediff/src/main.rs:11]: Is this part incomplete or to be removed?

File: utils/statediff/src/main.rs [hunk 3]
```diff
@@ -1,68 +1,86 @@
...
-enum Subcommand {
-    #[structopt(name = "chain")]
-    Chain(ChainCommand),
-}
+// /// statediff binary sub-commands available.
+// #[derive(Parser)]
+// enum Subcommand {
+//     #[command(name = "chain")]
+//     Chain(ChainCommand),
+// }
```
Comment (creativcoder) [inline utils/statediff/src/main.rs:53]: is this part incomplete or to be removed?

File: utils/statediff/src/main.rs [hunk 4]
```diff
@@ -1,68 +1,86 @@
...
+#[derive(Parser)]
 struct Cli {
-    #[structopt(subcommand)]
+    #[command(subcommand)]
     cmd: Subcommand,
 }
 
+#[derive(clap::Subcommand)]
+enum Subcommand {
+    /// does testing things
```
Comment (creativcoder) [inline utils/statediff/src/main.rs:64]: Maybe this comment can be improved

File: forest/cli/src/cli/mod.rs [hunk 5]
```diff
@@ -19,84 +19,73 @@ mod state_cmd;
...
+#[derive(clap::Subcommand)]
 pub enum Subcommand {
     /// Download parameters for generating and verifying proofs for given size
-    #[structopt(name = "fetch-params")]
+    #[command(name = "fetch-params")]
     Fetch(FetchCommands),
 
     /// Interact with Filecoin blockchain
-    Chain(ChainCommands),
+    Chain(ChainCommandsStruct),
```
Comment (hanabi1224) [inline forest/cli/src/cli/mod.rs:58]: Why use inconsistent pattern here between `Fetch` and the others? Any chance to remove the `*Struct` boilerplate code?

File: forest/shared/src/cli/mod.rs (hunk 1 above)
Comment (tyshko5) [reply forest/shared/src/cli/mod.rs:85]: Unfortunately, no

File: utils/statediff/src/main.rs (hunk 3 above)
Comment (tyshko5) [reply utils/statediff/src/main.rs]: Removed

File: forest/cli/src/cli/mod.rs (hunk 5 above)
Comment (tyshko5) [reply forest/cli/src/cli/mod.rs:58]: Unfortunately, in clap you cannot have subcommands that are enums of subcommands, therefore this workaround. That's the main difference between  `ChainCommands` and `Fetch`

File: utils/statediff/src/main.rs (hunk 2 above)
Comment (tyshko5) [reply utils/statediff/src/main.rs]: Removed

File: forest/cli/src/cli/mod.rs (hunk 5 above)
Comment (hanabi1224) [reply forest/cli/src/cli/mod.rs:58]: Thanks!

**Review by hanabi1224 — APPROVED** [review]

File: utils/statediff/src/main.rs (hunk 4 above)
Comment (tyshko5) [reply utils/statediff/src/main.rs]: It shouldn't be there at all. Removed

File: forest/cli/src/cli/mod.rs (hunk 5 above)
Comment (LesnyRumcajs) [reply forest/cli/src/cli/mod.rs:58]: @tyshko5 Why do you say you cannot have them? https://gist.github.com/LesnyRumcajs/dcd8be1f7bdcefcc438e2cdfa2da4a36

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
I don't really get the limitation you encountered, let's discuss it before any merge.

File: forest/cli/src/cli/mod.rs (hunk 5 above)
Comment (tyshko5) [reply forest/cli/src/cli/mod.rs]: Had an issue with this. Fixed it

File: utils/statediff/src/main.rs [hunk 6]
```diff
@@ -1,68 +1,63 @@
...
-}
-
 /// CLI structure generated when interacting with the statediff tool
-#[derive(StructOpt)]
-#[structopt(
-    name = env!("CARGO_PKG_NAME"),
-    version = env!("CARGO_PKG_VERSION"),
-    about = env!("CARGO_PKG_DESCRIPTION"),
-    author = env!("CARGO_PKG_AUTHORS")
-)]
```
Comment (LesnyRumcajs) [inline utils/statediff/src/main.rs:54]: Why did you delete it? The binaries lost, amongst others, versioning. Do you agree it's incorrect?

```
more_forest on  tyshko5/issue-2365-to-clap via 🦀 v1.69.0-nightly
❯ target/debug/forest_statediff --version
error: unexpected argument '--version' found

Usage: forest_statediff <COMMAND>

For more information, try '--help'.
```
vs
```
more_forest on  main via 🦀 v1.69.0-nightly took 40s
❯ target/debug/forest_statediff --version
forest_statediff 0.6.0

```

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

File: utils/statediff/src/main.rs (hunk 6 above)
Comment (tyshko5) [reply utils/statediff/src/main.rs]: Reverted to previous behaviour

Comment (LesnyRumcajs) [reply utils/statediff/src/main.rs]: @tyshko5 we still miss the author in the help.

structopt
```
❯ target/debug/forest_statediff --help
forest_statediff 0.6.0
ChainSafe Systems <info@chainsafe.io>
statediff binary sub-commands available

USAGE:
    forest_statediff <SUBCOMMAND>

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

SUBCOMMANDS:
    chain    Examine the state delta
    help     Prints this message or the help of the given subcommand(s)

```
clap (this branch)
```
❯ target/debug/forest_statediff --help

Usage: forest_statediff <COMMAND>

Commands:
  chain  
  help   Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version

```

Could you make it, so we don't lose any of that information? Otherwise, passing the author seems pointless if we don't use it anywhere. [This](https://docs.rs/clap/latest/clap/builder/struct.Command.html#method.help_template) may be needed.

Comment (tyshko5) [reply utils/statediff/src/main.rs]: The thing you've mentioned can only be used only through `Builder` pattern, and we are using `Derive`. Theoretically there are the same attributes [here](https://docs.rs/clap/latest/clap/_derive/index.html#command-attributes) , but they just  don't work. So I'd just delete those fields @LesnyRumcajs

Comment (tyshko5) [reply utils/statediff/src/main.rs:54]: Finally works as before @LesnyRumcajs

File: utils/statediff/Cargo.toml [hunk 7]
```diff
@@ -1,5 +1,7 @@
 [package]
 name = "forest_statediff"
+description = "statediff binary sub-commands available"
```
Comment (LesnyRumcajs) [inline utils/statediff/Cargo.toml:3]: The old description seemed odd; perhaps we can make it more like a description.
```suggestion
description = "Utility tool for debugging Filecoin state differences."
```

**Review by LesnyRumcajs — APPROVED** [review]

File: utils/statediff/Cargo.toml (hunk 7 above)
Comment (tyshko5) [reply utils/statediff/Cargo.toml]: Fixed

File: utils/statediff/src/main.rs [hunk 8]
```diff
@@ -1,67 +1,79 @@
...
+#[derive(Parser)]
+#[command(name = env!("CARGO_PKG_NAME"), author = env!("CARGO_PKG_AUTHORS"), version = env!("CARGO_PKG_VERSION"), about = env!("CARGO_PKG_DESCRIPTION"))]
+#[command(help_template(
+    "\
+{name} {version}
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply utils/statediff/src/main.rs]: Done

[the remark this answers is not in the document]
Comment (LesnyRumcajs) [reply utils/statediff/src/main.rs]: Fantastic!
